### ISSUE TITLE
Windows support

### ISSUE THREAD

**Issue by samschott** [samschott/desktop-notifier#1]
Windows support is currently still missing since I have no experience with Windows GUI frameworks.

Different options may be:

1. Use win32 and the `Shell_NotifyIconW` API documented [here](https://docs.microsoft.com/en-us/windows/win32/shell/notification-area#notifications-and-the-notification-area). Advantages: Can be used from pure Python code using `ctypes.windll`. See [plyer](https://github.com/kivy/plyer/blob/master/plyer/platforms/win/libs/balloontip.py) for an example implementation. Disadvantages: Does not appear to support buttons.
2. Use Toast Notification API as documented [here](https://docs.microsoft.com/en-us/windows/uwp/design/shell/tiles-and-notifications/toast-ux-guidance). Advantages: Simpler API. Richer UI with buttons etc. Disadvantages: Likely requires a package such as `pythonnet` with extension modules.

**Comment by MestreLion** [comment]
Might take a look at [Win10Toast](https://pypi.org/project/win10toast/), an implementation for the Toast Notification API that, as it seems, is pure python with absolutely zero dependencies other than pypiwin32 and setuptools (which I assume both are required in any Python for Windows software)

**Comment by SamSchott** [comment]
@MestreLion, thanks for the suggestion. From a brief look, it appears that `Win10Toast` depends on `pypiwin32` which is just a shim for [pywin32](https://github.com/mhammond/pywin32).  Pywin32 in turn requires compilation of extension modules. From their README:

> Building from source has been simplified recently - you just need Visual Studio and the Windows 10 SDK installed (the free compilers probably work too, but haven't been tested - let me know your experiences!)

They recommend to use their binary installer from the release page instead. This comes in at about 10 MB and I assume that it's a compressed format (though I have no idea how exe installers work). So it's not exactly purely Python and actually quite a heavy dependency.

In a worst case scenario, I'd be willing to go down the route of extension modules. Though we would probably want to write our own, to avoid unnecessary overhead from a general package like `pywin32`. My hope is that this can be avoided, but I don't have any experience with Windows...

**Comment by yusuf8ahmed** [comment]
Hi Sam,

I had the same idea as you and I have an implementation using [pywinrt](https://github.com/yusuf8ahmed/Creed/blob/new_development/creed/windows/xlang_toast.py) and [pythonnet](https://github.com/yusuf8ahmed/Creed/blob/new_development/creed/windows/pythonnet_toast.py) (i checked 1 month ago and it still working).

I would like to make a PR and include this in your package.

**Comment by SamSchott** [comment]
Hm, I was hoping to avoid libraries such as pywinrt or pythonnet. However, after reading the discussion of possible alternatives in https://github.com/beeware/toga/issues/884, I am not sure if this really is viable. So looking at the alternatives for bridging:

1) **ctypes access to the win32 API**: Gives us access to the C `Shell_NotifyIconW` API with limited functionality. However, the resulting notifications are displayed as toasts. This is similar to what [plyer](https://github.com/kivy/plyer/blob/master/plyer/platforms/win/libs/balloontip.py) and [Win10Toast](https://pypi.org/project/win10toast/) are doing. Win10Toast is using [pywin32](https://github.com/mhammond/pywin32) to do the bridging, a pre-compiled and large dependency. I'd prefer the ctypes approach which achieves the same effect. It will be more work to implement but will result in a much lighter package.

2) **WinRT**: By Microsoft but with apparently lacklustre maintenance. They do provide wheels for Python 3.7 - 3.9 on PyPI. `desktop_notifier` supports Python 3.6 to 3.9, so this would limit us.

3) **Python.net**: Seems to be more actively maintained now that they have funding by the .NET Foundation. Currently supports Python 3.5 to 3.8. Python 3.9 support seems to be coming soon, see https://github.com/pythonnet/pythonnet/pull/1264.

From those three options, my order of preference tends to be pythonnet then ctypes. The former introduces a tricky dependency but allows access to the more modern Toast API with richer notifications.

@yusuf8ahmed, what do you think? What is your experience with winrt vs pythonnet?

**Comment by yusuf8ahmed** [comment]
Hi @Sam,

Personally, I had a better experience with pythonnet. PYWinRT is still in its experimental phase and isn't mature enough to give to other developers. Nevertheless, we should keep an eye on it, as it is really promising and it might lead to a feature in the package where users can choose the backend that produces the notification

Also, I think it would be useful to merge your project into my project [Creed](https://github.com/yusuf8ahmed/Creed/tree/new_development) as it already has almost 6,000+ users?

**Comment by SamSchott** [comment]
> Personally, I had a better experience with pythonnet. PYWinRT is still in its experimental phase and isn't mature enough to give to other developers.

Good to know. My first impression is similar.

> Also, I think it would be useful to merge your project into my project Creed as it already has almost 6,000+ users?

I'm not sure how easy that would be, there seem to be a few fundamental differences between the projects:

1. The APIs are quite different. Creed appears to use the individual notification instance as a main interface while this package uses the "notification center" concept to allow global settings and notification management. Both are valid approaches, they are just different.
2. Platform implementations are handled differently between projects.
3. You are planning to offer multiple backends per platform while this library aims to offer a single backend only, ideally the most light-weight and versatile. This of course provides less choices for the user but makes it much easier to maintain. As long as the default implementation doesn't pose any installation problems (=> Python only) or reduced functionality, this should not limit use cases in any way.
4. Maturity: `desktop-notifier` may appear new but it's code is spun-off from a different project, `maestral`. As a result, the Linux and macOS backends already provide access to large parts of the platform APIs.

The spin-off from `maestral` has happened only in the development branch, but once it's official in the stable release, the user base of `desktop-notifier` will actually be quite large.

In any case, I'm happy to continue the discussion of possibly combing the projects, but preferably in a new issue.

**Comment by MestreLion** [comment]
> @MestreLion, thanks for the suggestion. From a brief look, it appears that `Win10Toast` depends on `pypiwin32` which is just a shim for [pywin32](https://github.com/mhammond/pywin32). Pywin32 in turn requires compilation of extension modules.

Yeah, I noticed that after a bit of research... and thank you the detailed response!

> Hm, I was hoping to avoid libraries such as pywinrt or pythonnet. (...)
>     1. **ctypes access to the win32 API**: Gives us access to the C `Shell_NotifyIconW` API with limited functionality. However, the resulting notifications _are_ displayed as toasts. (...)
>     3. **Python.net**: (...)
> From those three options, my order of preference tends to be pythonnet then ctypes. The former introduces a tricky dependency but allows access to the more modern Toast API with richer notifications.

Not sure if I'm entitled to an opinion (I don't even use Windows), but I'm also fond of a `ctypes` approach, even if you have to "strip down" `plyer` to the just the API calls/constants you need for a Notification. It's a lightweight, pure-python solution, relatively straightforward and easy to implement, and although the `Shell_NotifyIconW` API might be limited, as you said they _are_ displayed as Toasts in Win10+. You would at least have _some_ Windows support right away.

In the future, you _might_ replace this for a fully-fledged  Python.net / Toast Notification API. You could even initially add it as an `extras_require = {'toastapi': ['pythonnet']}` dependency in your `setup.py` and keep the "legacy" solution as a fallback, so the users could choose at pip-install time the tradeoff between "large compiled dependency/feature-rich API" or "lightweight pure-python/limited API"

**Comment by MestreLion** [comment]
> The spin-off from `maestral` has happened only in the development branch, but once it's official in the stable release, the user base of `desktop-notifier` will actually be quite large.
> 
> In any case, I'm happy to continue the discussion of possibly combing the projects, but preferably in a new issue.

Since we're already deviating a bit from the topic, let me take some time to congratulate you: I've recently stumbled on `maestral` several times the last few weeks, when looking for solutions/issues/inspiration regarding cross-platform XDG paths, notifications, keyrings, etc (that design flaw in python-keyring is SO annoying!).

And in every single issue the decision and approach taken by `maestral` (and to an extent `desktop-notifier`) was the most elegant and sensible I've seen. Choosing the DE's default keyring to avoid kwallet GUI popping up in Gnome, respect `$XDG_CONFIG_HOME` if set on MacOS, auto-choosing the best notification backend (and using `Dbus` directly on Linux), all great design decisions. Kudos for you, @SamSchott !

**Comment by SamSchott** [comment]
> Not sure if I'm entitled to an opinion (I don't even use Windows), but I'm also fond of a ctypes approach

Of course you are entitled to an opinion :) Yes, I've tried to keep dependencies very simple, with Objective-C calls handled via ctypes / [rubicon-objc](https://github.com/beeware/rubicon-objc) and dbus communication via [dbus-next](https://github.com/altdesktop/python-dbus-next) which is also purely Python. But it looks like there really is no C API to the Windows .NET runtime.

In any case, I'll wait for some input from actual Windows users (which I am not), before making a call. And if a Windows user would like to contribute, they are very welcome! A first implementation doesn't need to support all features.

> And in every single issue the decision and approach taken by maestral (and to an extent desktop-notifier) was the most elegant and sensible I've seen. 

This is good to hear! A lot of the choices really were made after user feedback, the GitHub and Linux crowd are very quick to complain once something is *slightly* off. And I have learned what a joy it is to write for the macOS platform compared to the fragmented Linux desktop ecosystem...
